README

ExtUtils::Git is a class that extends ExtUtils::MakeMaker to add git related
targets to the Makefile generated from Makefile.PL.

ExtUtils::Git will enforce various rules during module distribution build,
such as not building a dist for a module before all components are checked
in to Git.  It will also not build a dist if the MANIFEST and Git ideas of
what are in the module are out of sync.

ExtUtils::Git will take care of incrementing the version number of modules
before a dist is built and other such mundane details.

ChangeLog
2007-07-01 19:16  aspeer

	* VERSION.pm (1.55): ExtUtils-CVS version increment

2005-11-07 11:13  aspeer

	* VERSION.pm (1.54): ExtUtils-CVS version increment

2005-10-23 16:55  aspeer

	* VERSION.pm (1.53): ExtUtils-CVS version increment

2005-02-06 13:01  aspeer

	* VERSION.pm (1.52): ExtUtils-CVS version increment

2005-02-04 22:07  aspeer

	* CVS.pm (1.36), Config.pm (1.16): Added NO_META pragma to
	  MakeMaker files, fixed makemaker section to rebuild Makefile.PL
	  after change checkin etc to work with new ExtUtils::MakeMaker

2005-02-04 22:07  aspeer

	* VERSION.pm (1.51): ExtUtils-CVS version increment

2004-06-18 01:52  aspeer

	* VERSION.pm (1.50): ExtUtils-CVS version increment

2004-06-18 01:52  aspeer

	* CVS.pm (1.35): Added lame errnofatal import target - not needed,
	  delete later

2004-06-17 23:08  aspeer

	* VERSION.pm (1.49): ExtUtils-CVS version increment

2004-06-17 23:08  aspeer

	* CVS.pm (1.34): Modify ci_status_bundle to supply file mtime when
	  doint mtime_sync, fix some error messages to clear up file name
	  in error

2004-06-17 22:19  aspeer

	* VERSION.pm (1.48): ExtUtils-CVS version increment

2004-06-17 22:19  aspeer

	* CVS.pm (1.33): Removed debugging info, fixed formatting

2004-06-08 11:05  aspeer

	* VERSION.pm (1.47): ExtUtils-CVS version increment

2004-03-27 17:25  aspeer

	* VERSION.pm (1.46): ExtUtils-CVS version increment

2004-02-06 01:26  aspeer

	* VERSION.pm (1.45): ExtUtils-CVS version increment

2004-02-06 01:26  aspeer

	* CVS.pm (1.30), Config.pm (1.14), support/dist_ci.inc (1.21):
	  Convert private utility functions to OO methods so can be
	  inherited by other ExtUtils:: modules

2004-02-05 23:10  aspeer

	* VERSION.pm (1.44): ExtUtils-CVS version increment

2004-02-05 23:10  aspeer

	* CVS.pm (1.29), Config.pm (1.13), support/dist_ci.inc (1.20):
	  Change name and format of Dumper file tp '.module.pm'. Revert
	  back to old way of dumping version in dist_ci.

2004-02-05 22:01  aspeer

	* VERSION.pm (1.43): ExtUtils-CVS version increment

2004-02-05 22:01  aspeer

	* CVS.pm (1.28), support/dist_ci.inc (1.19): Interim checkin -
	  convert method params from array to hash ref, experiment with
	  version dump

2004-02-05 16:40  aspeer

	* VERSION.pm (1.42): ExtUtils-CVS version increment

2004-02-05 16:40  aspeer

	* CVS.pm (1.27), Config.pm (1.12), support/dist_ci.inc (1.18):
	  Added routines to modify Makemaker "makefile" target so that
	  Makefile is rebuilt correctly after checkin, should rebuild with
	  exact ExtUtils modules used to create in the first place, eg if
	  called thus:

	  perl -MExtUtils::CVS=:all Makefile.PL ...  make ci

	  Then Makefile.PL will be called with exact same args

2004-02-04 17:22  aspeer

	* VERSION.pm (1.41): ExtUtils-CVS version increment

2004-02-04 17:22  aspeer

	* support/dist_ci.inc (1.17): Change version increment message in
	  ChangeLog

2004-02-04 17:17  aspeer

	* VERSION.pm (1.40): Version increment

2004-02-04 17:17  aspeer

	* CVS.pm (1.26), Config.pm (1.11), support/dist_ci.inc (1.16): Get
	  rid of obselete ci_changelog target

2004-02-04 17:15  aspeer

	* VERSION.pm (1.39): Version increment

2004-02-04 17:15  aspeer

	* Config.pm (1.10): Next attempt ..

2004-02-04 17:14  aspeer

	* VERSION.pm (1.38): Version increment

2004-02-04 17:14  aspeer

	* CVS.pm (1.25), Config.pm (1.9), support/dist_ci.inc (1.15): Next
	  attempt at versionin ChangeLog

2004-02-04 16:49  aspeer

	* CVS.pm (1.24): Further attempt at getting overall module version
	  into ChangeLog

2004-02-04 16:38  aspeer

	* CVS.pm (1.23), Config.pm (1.8), support/dist_ci.inc (1.14):
	  Update to attempt including version number into ChangeLog

2004-02-04 15:40  aspeer

	* VERSION.pm (1.35): Version increment

2004-02-04 15:40  aspeer

	* CVS.pm (1.22): Code tidyup

2004-02-04 15:39  aspeer

	* VERSION.pm (1.34): Version increment

2004-02-04 15:39  aspeer

	* Config.pm (1.7): Altered CVS2CL command line args - added unify
	  window time of 15 secs, resorted to using --file intead of -F to
	  get around small bug

2004-02-04 15:07  aspeer

	* VERSION.pm (1.33): Version increment

2004-02-04 15:07  aspeer

	* CVS.pm (1.21), Config.pm (1.6), MANIFEST (1.3), Makefile.PL
	  (1.5), support/cvs2cl.pl (1.1), support/dist_ci.inc (1.13): Big
	  rewrite to attempt streamlining and clarification of code.
	  Separated code into sections of public/private methods.

	  Re-wrote dist_ci target to be clearer

	  Added options in Config file to get better output from cvs2cl.pl
	  file, which is now included in the distribution

2003-11-03 12:42  aspeer

	* CVS.pm (1.20): Change mtime_sync alorithm again, attemp to
	  support DST

2003-10-16 00:28  aspeer

	* CVS.pm (1.18), Config.pm (1.4), support/dist_ci.inc (1.11): First
	  crack at Win32 compatible code for cvs manipulation of files

2003-09-01 16:51  aspeer

	* CVS.pm (1.16), Makefile.PL (1.3): Added pod doc to CVS.pm,
	  removed empty man3pods sections from Makefile.PL

2003-09-01 14:05  aspeer

	* CVS.pm (1.15), Config.pm (1.2), LICENSE (1.1), MANIFEST (1.2),
	  Makefile.PL (1.2), VERSION.pm (1.24): Ran through ExtUtils::Lint,
	  added license and copyright info into module

2003-07-06 18:06  aspeer

	* CVS.pm (1.14): Added routine in "manicheck" method to check for
	  patch dir, and if found make sure all patch files are in CVS and
	  the MANIFEST

2003-06-17 00:13  aspeer

	* CVS.pm (1.13): Remove File::Touch module, add reminder if we use
	  it again

2003-06-15 16:14  aspeer

	* CVS.pm (1.12), support/dist_ci.inc (1.10): No longer 'touch'
	  files that have mtime out of sync with CVS status, allow 'cvs
	  status' to do this for us. Modify the may we go through the CVS
	  'Entries' file, read it all in at once and iterate. Previously
	  going through it line by line, but were doing 'cvs' operations
	  that changed it while we were iterating - result was missed some
	  files, did some needlessly

	  Update dist_ci to do 'cvs status' before check-in to update file

Config.pm
#
#
#  Copyright (c) 2003 Andrew W. Speer <andrew.speer@isolutions.com.au>. All rights 
#  reserved.
#
#  This file is part of ExtUtils::CVS.
#
#  ExtUtils::CVS is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
#  $Id: Config.pm,v 1.18 2007/07/01 09:46:09 aspeer Exp $
#


#  This package finds, then stores in a cache the full path to various
#  executables and environmnet vars, including:
#
#  cvs
#  cvs2cl
#  CVSROOT
#
#  If no cache file is found, this modules will go through the path and
#  look for the executables, then store the results away in a cache file
#
#


#
#
package ExtUtils::CVS::Config;


#  Compiler Pragma
#
sub BEGIN   { $^W=0 };
use strict  qw(vars);
use warnings;
no  warnings qw(uninitialized);


#  Need File::Find, other File utils
#
use File::Find;
use File::Spec;
use IO::File;


#===================================================================================================


#  Bin_find now anon sub to stop "subroutine redefined" type errors that
#  occur when this file is read several times
#
my $bin_find_cr=sub {


    #  Find a binary file
    #
    my $bin_ar=shift();
    my $bin_fn;


    #  Find the bin file/files if given array ref. If not supplied as array ref
    #  convert.
    #
    (ref($bin_ar) eq 'ARRAY') || do { $bin_ar=[$bin_ar] };
    my $wanted_cr=sub {
	foreach my $bin (@{$bin_ar}) {
	    ($File::Find::name=~/\/\Q$bin\E$/) && ($bin_fn=$File::Find::name);
	    last if $bin_fn;
	}
    };
    my @dir=grep { -d $_ } split(/:|;/, $ENV{'PATH'});
    find($wanted_cr, @dir);


    #  Normalize fn
    #
    $bin_fn=File::Spec->canonpath($bin_fn);


    #  Return
    #
    return $bin_fn;

};


#  Get cache file name
#
my $cache_fn=$INC{'ExtUtils/CVS.pm'} || File::Spec->rel2abs(__FILE__);
$cache_fn .= '.cache';


#  Cache can only be 60 sec old
#
if (-e $cache_fn && ((stat($cache_fn))[9] < (time() - (1 * 60)))) {


    #  Cache is stale, delete
    #
    unlink $cache_fn ||
	die("unable to unlink $cache_fn, $!");

}


#  Try to read in cache details, or search disk for binaries if needed
#
unless ($_ = do($cache_fn)) {


    #  Could not find cache file, create hash. Should probably put this stuff into a
    #  support/const.inc file later as a template, as contains some constants
    #
    my %Config= (

	CVS			 =>  $bin_find_cr->([qw(cvs cvs.exe)]),

	CVS2CL			 =>  $bin_find_cr->('cvs2cl.pl'),

	CVSROOT			 =>  $ENV{'CVSROOT'},

	CHANGELOG		 =>  $_='ChangeLog',

	CVS2CL_ARG		 =>  "--window 15 --file $_ -P -r -I $_",
	
	METAFILE		 =>  'META.yml',

	DUMPER_FN	         =>  '.dumper.pm',

	EXTUTILS_ARGV		 =>  q["$(NAME)" "$(NAME_SYM)" "$(DISTNAME)" "$(DISTVNAME)" "$(VERSION)" ].
	    q["$(VERSION_SYM)" "$(VERSION_FROM)"],

	EXTUTILS_CVS		 =>  'ExtUtils::CVS',
	
       );


    #  Store in cache file. Does not matter if not writeable
    #
    if (my $fh=IO::File->new($cache_fn, O_WRONLY|O_CREAT|O_TRUNC)) {
        print $fh &Data::Dumper::Dumper(\%Config)
    }


    #  Return ref
    #
    $_=\%Config;

}
